src/github-adapter.ts

```diff
@@ -0,0 +1,143 @@
+const moment = require('moment');
+
+import fetch from 'node-fetch';
+import * as IDataAdapter from './data-adapter';
+import * as ICacheAdapter from './cache-adapter';
+import sequence from './utils/promise-sequence';
+
+interface IGithubPaginationRels {
+    [key: string]: string;
+}
+
+interface IResponseWrapper {
+    rels: IGithubPaginationRels;
+    response: any;
+}
+
+export default class GithubAdapter implements IDataAdapter.IDataAdapterInterface {
+    private token: string;
+    private username: string;
+    private organizationName: string;
+    private cacheAdapter: ICacheAdapter.ICacheAdapterInterface;
+
+    constructor(token: string, username: string, organizationName: string, cacheAdapter: ICacheAdapter.ICacheAdapterInterface) {
+        this.token = token;
+        this.username = username;
+        this.organizationName = organizationName;
+        this.cacheAdapter = cacheAdapter;
+    }
+
+    fetchData(fromDate: Date, endDate: Date): Promise<any> {
+        return this.cacheAdapter.readCache().catch((err) => {
+            if (!(err instanceof ICacheAdapter.CacheEmptyError)) throw err;
+
+            const items : Array<any> = [];
+
+            const parsedFromDate : string = moment(fromDate).format('YYYY-MM-DD');
+            const parsedEndDate : string = moment(endDate).format('YYYY-MM-DD');
+
+            const firstHref = `https://api.github.com/search/issues?per_page=50&q=type:pr org:${this.organizationName} merged:${parsedFromDate}..${parsedEndDate}`;
+
+            return this.fetchOnePage(firstHref, items).then(() => {
+                console.log('Fetched ' + items.length + ' items');
+                const uniqueItems = this.extractUniqueItems(items);
+                const pullRequests : Array<Promise<any>> = uniqueItems.map(item => this.fetchPullRequest(item));
+                const results : any = [];
+                return this.fetchDataInSequenceToPreventGithubAbuse(pullRequests, results);
+            }).then((items : Array<any>) => {
+                const pullRequestsWithReviews : Array<Promise<any>> = items.map(item => this.addReviewsToPullRequests(item));
+                const results : any = [];
+                return this.fetchDataInSequenceToPreventGithubAbuse(pullRequestsWithReviews, results);
+            }).then((items) => {
+                this.cacheAdapter.writeCache(items);
+                return Promise.resolve(items);
+            });
+        });
+    }
+
+    private fetchDataInSequenceToPreventGithubAbuse(tasks : Array<any>, results : Array<any>) : Promise<any> {
+        return sequence(tasks, results);
+    }
+
+    private fetchPullRequest(pullRequest : any) : any {
+        return () => {
```

Comment:
Because I'm running sequence of tasks generated by this HOF - https://github.com/jniechcial/code-review-watchman/pull/1/files#diff-9857a48cb5c77212c8810513f272a226R59. They are executed synchronously one by one.